packages/@aws-cdk/aws-stepfunctions-tasks/lib/dynamodb/shared-types.ts

```diff
@@ -0,0 +1,189 @@
+/**
+ * Determines the level of detail about provisioned throughput consumption that is returned.
+ */
+export enum DynamoConsumedCapacity {
+  /**
+   * The response includes the aggregate ConsumedCapacity for the operation,
+   * together with ConsumedCapacity for each table and secondary index that was accessed
+   */
+  INDEXES = 'INDEXES',
+
+  /**
+   * The response includes only the aggregate ConsumedCapacity for the operation.
+   */
+  TOTAL = 'TOTAL',
+
+  /**
+   * No ConsumedCapacity details are included in the response.
+   */
+  NONE = 'NONE',
+}
+
+/**
+ * Determines whether item collection metrics are returned.
+ */
+export enum DynamoItemCollectionMetrics {
+  /**
+   * If set to SIZE, the response includes statistics about item collections,
+   * if any, that were modified during the operation.
+   */
+  SIZE = 'SIZE',
+
+  /**
+   * If set to NONE, no statistics are returned.
+   */
+  NONE = 'NONE',
+}
+
+/**
+ * Use ReturnValues if you want to get the item attributes as they appear before or after they are changed
+ */
+export enum DynamoReturnValues {
+  /**
+   * Nothing is returned
+   */
+  NONE = 'NONE',
+
+  /**
+   * Returns all of the attributes of the item
+   */
+  ALL_OLD = 'ALL_OLD',
+
+  /**
+   * Returns only the updated attributes
+   */
+  UPDATED_OLD = 'UPDATED_OLD',
+
+  /**
+   * Returns all of the attributes of the item
+   */
+  ALL_NEW = 'ALL_NEW',
+
+  /**
+   * Returns only the updated attributes
+   */
+  UPDATED_NEW = 'UPDATED_NEW',
+}
+
+/**
+ * Class to generate projection expression
+ */
+export class DynamoProjectionExpression {
+  private expression: string[] = [];
+
+  /**
+   * Adds the passed attribute to the chain
+   *
+   * @param attr Attribute name
+   */
+  public withAttribute(attr: string): DynamoProjectionExpression {
+    if (this.expression.length) {
+      this.expression.push(`.${attr}`);
+    } else {
+      this.expression.push(attr);
+    }
+    return this;
+  }
+
+  /**
+   * Adds the array literal access for passed index
+   *
+   * @param index array index
+   */
+  public atIndex(index: number): DynamoProjectionExpression {
+    if (!this.expression.length) {
+      throw new Error('Expression must start with an attribute');
+    }
+
+    this.expression.push(`[${index}]`);
+    return this;
+  }
+
+  /**
+   * converts and return the string expression
+   */
+  public toString(): string {
+    return this.expression.join('');
+  }
+}
+
+/**
+ * Class to generate AttributeValue
+ */
+export interface DynamoAttributeValue {
```

Comment:
This API is gross. It allows users to happily craft invalid inputs such as:
- `{}`
- `{ s: 'foo', n: 'bar' }`

I'm certain we can do better than this for our customers.

Reply:
agreed. it did have some validation but allowing invalid construction itself is worth addressing.